### PR TITLE
Phase 6: Reynolds-Conditional SRF — FiLM on (Re, AoA) for Surface Refinement Head

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The shared `surface_refine` head (`--surface_refine`, merged Phase 5, the single biggest win at -72.7% p_re) currently applies the same MLP correction regardless of flow regime. It has no awareness of Reynolds number or angle of attack — the same weights correct predictions at Re=1e5 and Re=4e6, even though the error character differs dramatically:

- **Low Re:** boundary layer separation → large Cp plateaus, smooth gradients
- **High Re:** attached flow → sharp leading-edge gradients, pressure recovery

**Hypothesis:** Conditioning the SRF head on (Re, AoA) via FiLM modulation allows the correction MLP to adapt its strategy to the flow regime, directly targeting `p_re` and `p_oodc`.

**Why this is different from the dead-end `aft_foil_srf_film` (+41.6% p_oodc regression, #2104):** That FiLM used configurational features (gap, stagger) that vary per-sample dynamically. Re and AoA are the *primary flow condition descriptors* — they are already used by AdaLN in every Transolver block. FiLM on Re/AoA for the SRF head is the same mechanism as AdaLN, applied to the post-processing correction step.

**Expected impact:** -2 to -4% p_re, -1 to -2% p_oodc. ~18 LoC change. No memory overhead.

### Step 1: Locate the key classes

```bash
grep -n "class SurfaceRefinementHead\|class AftFoilRefinementHead\|film" cfd_tandemfoil/train.py | head -30
```

Find `SurfaceRefinementHead.__init__` and `SurfaceRefinementHead.forward`. Also look at `AftFoilRefinementHead` — it already has `film=True` FiLM support; mirror that exact pattern.

### Step 2: Add FiLM to SurfaceRefinementHead

In `SurfaceRefinementHead.__init__`, add optional `film: bool = False` parameter and FiLM layers:

```python
self.film = film
if film:
    self.film_scale = nn.Linear(2, hidden_dim, bias=False)
    self.film_shift = nn.Linear(2, hidden_dim)
    nn.init.zeros_(self.film_scale.weight)
    nn.init.zeros_(self.film_shift.weight)
    nn.init.zeros_(self.film_shift.bias)
```

In `SurfaceRefinementHead.forward`, add a `cond` argument and apply FiLM after the first hidden activation:

```python
def forward(self, h_surf, base_surf, cond=None):
    x = torch.cat([h_surf, base_surf], dim=-1)
    x = self.act(self.fc1(x))
    if self.film and cond is not None:
        scale = self.film_scale(cond)
        shift = self.film_shift(cond)
        x = x * (1 + scale) + shift
    x = self.fc2(x)
    return x
```

### Step 3: Verify the Re and AoA feature indices

```bash
grep -n "adaln\|cond_dim\|cond_feat\|cond\[:, 0\]\|x\[:, 0,\|condition" cfd_tandemfoil/train.py | head -40
```

The condition input should be `x[:, 0, 13:15]` — Re and AoA as the first 2 scalar condition features (same as what AdaLN uses). Verify by printing at training start:

```python
print(f"[DEBUG] cond sample: feat13={x[0,0,13]:.4f}, feat14={x[0,0,14]:.4f}")
```

These should be normalized values in roughly [-3, 3]. If they look like zeros, find the right indices via `prepare_multi.py` or the stats dict.

### Step 4: Pass condition to the SRF head

Find where `refine_head(` is called. Add condition extraction and pass it:

```python
# Extract per-sample condition [B, 2]
cond_global = x[:, 0, 13:15]

# Expand to surface nodes using surf_idx (surf_idx[:, 0] = batch index per node)
cond_surf = cond_global[surf_idx[:, 0]]   # [N_surf_nodes, 2]

# Call with condition:
surf_pred = refine_head(h_surf, base_surf, cond=cond_surf)
```

### Step 5: Add config flags

```python
# argparse:
parser.add_argument('--surface_refine_film', action='store_true')
# Config dataclass:
surface_refine_film: bool = False
```

Pass `film=cfg.surface_refine_film` when constructing `SurfaceRefinementHead`.

### Step 6: Run experiments

Run 2 film seeds + 2 control seeds (identical config, no `--surface_refine_film`) for a clean comparison:

```bash
# Film, seed 42
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/re-srf-film-s42" \
  --wandb_group phase6/re-conditional-srf \
  --surface_refine --surface_refine_film --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02

# Film, seed 73 — same + --seed 73
# Control, seed 42 — same without --surface_refine_film
# Control, seed 73 — same without --surface_refine_film + --seed 73
```

**Note:** Use `--surface_refine` (not `--surface_refine_hidden`/`--surface_refine_layers` alone) — the `--surface_refine` flag enables the head.

**W&B group:** `phase6/re-conditional-srf`

Report surface MAE metrics (p_in, p_oodc, p_tan, p_re) for all 4 runs. If both film seeds beat both control seeds on p_re, strong candidate for 8-seed validation.

## Baseline

Current single-model baseline (PR #2104 + #2115, +aft_foil_srf +aug_gap_stagger_sigma=0.02, 8-seed mean, seeds 42-49):

| Metric | 8-seed mean | Target to beat |
|--------|-------------|----------------|
| p_in   | **13.19 ± 0.33** | < 13.19 |
| p_oodc | **7.92 ± 0.17**  | < 7.92  |
| p_tan  | **30.05 ± 0.36** | < 30.05 |
| p_re   | **6.45 ± 0.07**  | **< 6.45** ← primary target |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/baseline-s42" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Reynolds-Conditional SRF (FiLM)

### Setup
- 4 runs: 2 FiLM + 2 control, seeds 42 and 73
- All runs: `--ema_start_epoch 100 --ema_decay 0.999 --cosine_T_max 160 --aft_foil_srf --aug_gap_stagger_sigma 0.02 --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 --surface_refine_hidden 192 --surface_refine_layers 3 --residual_prediction`
- FiLM runs add: `--surface_refine_film` (FiLM conditioning on (Re, AoA) in SurfaceRefinementHead)
- Training cut at 180-min timeout. Control reached epoch 151, FiLM reached epoch 149.
- EMA active from epoch 100 → ~51 epochs (ctrl) / ~49 epochs (film) of EMA accumulation.

### Results — Surface MAE (Pressure)

| Run | p_in | p_oodc | p_tan | p_re | W&B Run ID |
|-----|------|--------|-------|------|------------|
| **Control s42** | 12.8 | 7.8 | 29.7 | 6.8 | `np3anmah` |
| **Control s73** | 14.2 | 7.3 | 29.9 | 6.5 | `fcri663y` |
| **Control avg** | **13.5** | **7.55** | **29.8** | **6.65** | — |
| **FiLM s42** | 13.3 | 8.0 | 30.6 | 6.6 | `7bnlke5o` |
| **FiLM s73** | 12.7 | 7.7 | 29.7 | 6.5 | `yh5ixlx8` |
| **FiLM avg** | **13.0** | **7.85** | **30.15** | **6.55** | — |
| **Baseline (8-seed)** | **13.19** | **7.92** | **30.05** | **6.45** | — |

### FiLM vs Control (2-seed averages)

| Metric | FiLM | Control | Delta | Interpretation |
|--------|------|---------|-------|----------------|
| p_in | 13.0 | 13.5 | -3.7% | FiLM slightly better |
| p_oodc | 7.85 | 7.55 | +4.0% | FiLM slightly worse |
| p_tan | 30.15 | 29.8 | +1.2% | FiLM slightly worse |
| p_re | 6.55 | 6.65 | -1.5% | FiLM slightly better |

### Analysis

**Null result.** FiLM conditioning on (Re, AoA) for the Surface Refinement Head shows no meaningful improvement over the control. The differences between FiLM and control are within seed noise — the single-seed variation (ctrl-s42 p_in=12.8 vs ctrl-s73 p_in=14.2 is a 10% swing) dwarfs the FiLM effect.

**Why it likely doesn't help:**
1. The SurfaceRefinementHead already receives (Re, AoA) implicitly through the hidden states from the Transolver backbone, which uses AdaLN conditioning on these same features at every block.
2. The refinement head's correction is a small residual — the heavy lifting is done by the backbone which already adapts to Re/AoA via AdaLN.
3. The FiLM layer only modulates the first hidden activation (after LayerNorm+GELU), so its effect is limited to a single layer of a 3-layer MLP.
4. FiLM adds only 960 extra parameters — too few to capture meaningful Re-dependent correction patterns beyond what the backbone already provides.

**Note on control runs:** Both control averages are within normal baseline variation. The `--ema_start_epoch 100` (vs baseline's default) may explain minor deviations.

### Suggested Follow-ups
1. **Skip FiLM for SRF.** The backbone's AdaLN already handles Re/AoA conditioning well enough. Adding explicit conditioning at the SRF level is redundant.
2. **Focus on loss formulation or data representation** rather than more conditioning mechanisms — the model's capacity to condition on flow regime is already well-served by AdaLN.
3. **Consider flow topology-aware refinement** — instead of conditioning on scalar Re/AoA, condition on local flow features (e.g., pressure gradient magnitude, boundary layer indicators) that vary spatially and aren't already captured by the global conditioning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2128 because Null result — W&B-verified FiLM 2-seed avg: p_in=13.03, p_oodc=7.84, p_tan=30.15, p_re=6.56. All metrics at or worse than baseline (p_oodc +2.4%, p_tan +0.1%, p_re +0.8%). FiLM is also worse than its own control on p_oodc (7.84 vs 7.55) and p_tan (30.15 vs 29.78).

Student's analysis is correct: the SRF head already receives Re/AoA conditioning implicitly through the Transolver backbone's AdaLN, so adding explicit FiLM modulation is redundant. The 960-parameter FiLM layer cannot learn anything new that AdaLN hasn't already encoded.

Good experiment design and thorough analysis. The follow-up suggestion (local flow feature conditioning rather than global scalars) is an interesting direction — something like conditioning on local pressure gradient magnitude or boundary layer thickness proxies rather than global (Re, AoA) might circumvent the redundancy problem..
